Added trigger to dataset table to update history.update_time

Nerdinacan opened the pull request:
This is a database trigger that updates the history table (updated_time) whenever a dataset changes. Currently, modifications to a dataset belonging to a history do not necessarily propagate upwards to the history or history_dataset_association table.

Having an accurate update_time on the history object will allow the client to more intelligently poll for updates. This should save a lot of querying from the client.

Alternative ideas I threw out include:

1. Having a view, or equivalently an ORM relationship object which does a subquery to generate the current update_time via max(update_time) on a 2-table subquery join. This seems like a performance sink since we'll be spamming this update_date every couple of seconds looking for updates until such time as we have a push-tech like a websocket.

2. Sifting through the server-side code to each place we update a dataset and modify it to do the equivalent work of the trigger. This could still work, but on the whole triggers will be more performant and maintaining timestamps is probably the best use of a trigger. (Except maybe audit trails.)

PostgreSQL has a proprietary trigger format that requires the installation of a database function which is referenced by the trigger, so I have split the migration according to postgres/non-postres.

I tested the installation on the default SQLite installation that worked fine. I also attempted to test against MySQL only to find that our migration scripts fail against MySQL way before they even get to this update, so I guess we're not super-concerned about that SQL variant.

Nerdinacan committed:
Added trigger to dataset table to update history.update_time

Nerdinacan committed:
Updated sql for finicky sqlite

Nerdinacan committed:
Slave to the linting overlords

nsoranzo commented on lib/galaxy/model/migrate/versions/0155_add_content_update_time.py:
```diff
@@ -0,0 +1,105 @@
+        CREATE TRIGGER BEFORE_{operation}_DATASET
+            BEFORE {operation} ON dataset
+            BEGIN
+                update history
```
Can you use uppercase for the SQL keywords also here and below?

Nerdinacan replied:
Sure, I'm not picky.

nsoranzo commented on lib/galaxy/model/migrate/versions/0155_add_content_update_time.py:
```diff
@@ -0,0 +1,105 @@
+def build_trigger(op):
+    create_trigger_template = """
+        CREATE TRIGGER BEFORE_{operation}_DATASET
+            BEFORE {operation} ON dataset
```
Why not
```suggestion
            AFTER {operation} ON dataset
```
? It makes more sense to me to update the history timestamp after the change to its content.

Nerdinacan replied:
Sure, I guess an error could get thrown or something. After might be more appropriate.

jdavcs replied:
@Nerdinacan @nsoranzo aren't all these supposed to be `AFTER`? Current code uses `BEFORE`. Unless there's been more discussion and it's been agreed to use `BEFORE`, in which case please ignore this comment.

nsoranzo replied:
@ic4f It was change back to `BEFORE` by @Nerdinacan in commit 335499fc5449dfd7ff38104ac84f5e522220425c , not sure why.

jdavcs replied:
Oh, OK, thanks for checking @nsoranzo

nsoranzo commented on lib/galaxy/model/migrate/versions/0155_add_content_update_time.py:
```diff
@@ -0,0 +1,105 @@
+        end;
+        $BODY$;
+        CREATE TRIGGER update_history_update_time
+            BEFORE INSERT OR DELETE OR UPDATE
```
```suggestion
            AFTER INSERT OR DELETE OR UPDATE
```
See below for explanation.

nsoranzo commented on lib/galaxy/model/migrate/versions/0155_add_content_update_time.py:
```diff
@@ -0,0 +1,105 @@
+            RETURNS trigger
+            LANGUAGE 'plpgsql'
+        AS $BODY$
+        begin
```
Can you use uppercase for SQL keywords here and below?

Nerdinacan replied:
Yep.

nsoranzo commented on lib/galaxy/model/migrate/versions/0155_add_content_update_time.py:
```diff
@@ -0,0 +1,105 @@
+            ON dataset
+            FOR EACH ROW
+            EXECUTE PROCEDURE update_history_content_update_time();
+    """).execute_if(dialect='postgresql')
```
Wouldn't be easier (and more consistent with other migrations) to use:
```python
        if migrate_engine.name in ['postgres', 'postgresql']:
            DDL(...).execute(bind=migrate_engine)
        else:
            ...
```
instead of the various `execute_if()`s ?

Nerdinacan replied:
Ahhh sure, was just using the API I found in the docs, but if you have a preferred way of doing it, I'm easy.

Nerdinacan committed:
added linting and formatting suggestions to trigger generation sql

Nerdinacan committed:
fix sql syntax error in drop statement

nsoranzo commented on lib/galaxy/model/migrate/versions/0155_add_content_update_time.py:
```diff
@@ -0,0 +1,102 @@
+def build_trigger(op):
+    create_trigger_template = """
+        CREATE TRIGGER AFTER_{operation}_DATASET
+            AFTER {operation} ON dataset
```
```suggestion
            AFTER {operation}
            ON dataset
            FOR EACH ROW
```

`FOR EACH ROW` is required by MySQL and optional on SQLite. According to [SQLite docs](https://www.sqlite.org/lang_createtrigger.html), "at this time SQLite supports only FOR EACH ROW triggers, not FOR EACH STATEMENT triggers", so it may be added in the future and being explicit seems safer.

nsoranzo commented on lib/galaxy/model/migrate/versions/0155_add_content_update_time.py:
```diff
@@ -0,0 +1,102 @@
+                RETURN NEW;
+            END;
+            $BODY$;
+            CREATE TRIGGER update_history_update_time
```
I was thinking we may want to establish a trigger naming convention, what about `trigger_TABLE_OPERATION`? For this trigger that would be `trigger_dataset_aiudr`, where `aiudr` stands for "after insert, delete of update for each row".

Nerdinacan replied:
Okie dokie.

nsoranzo commented on lib/galaxy/model/migrate/versions/0155_add_content_update_time.py:
```diff
@@ -0,0 +1,102 @@
+
+def build_trigger(op):
+    create_trigger_template = """
+        CREATE TRIGGER AFTER_{operation}_DATASET
```
Here the trigger name could be `trigger_dataset_a{op_initial}r`, where you set `op_initial` to `op.lower()[0]`

Nerdinacan replied:
Sure. I've renamed all the triggers to that format.

nsoranzo reviewed (commented):
Thanks for the changes, a couple more suggestions!

nsoranzo commented:
Ah, I also have a question: what about changes to the `history_dataset_association` table, do we need another trigger to update `history` for those too?

Nerdinacan commented:
@nsoranzo  I don't think I need a trigger on that many-many table. I'm already hearing some changes, likely through the ORM. It's the changes directly to the referenced dataset which are not propagating upwards.

I'm not against adding more update_time propagation triggers if it turns out that I DO need it in the future though. But that would not solve any immediate problems for me at this time.

Nerdinacan committed:
Update lib/galaxy/model/migrate/versions/0155_add_content_update_time.py

Co-Authored-By: Nicola Soranzo <nicola.soranzo@gmail.com>

Nerdinacan committed:
Modified trigger names to conform to requested format

Nerdinacan commented:
@nsoranzo  I may have to eat my words. 

I did, in fact, just run into a situation that might necessitate another trigger on the HDA table (or an equivalent ORM update).

Is there a way to put this PR on "hold" so nobody merges it until I get a chance to look more carefully? Or should I just close it?

Nerdinacan commented:
Thanks!

nsoranzo commented:
Maybe the trigger on dataset could update history_dataset_association and another trigger on history_dataset_association could update history?

Let us know when it's ready for review again.

Nerdinacan commented:
Yeah we could do a full chained propagation like that. @mvdbeek also suggested hooking into sql-alchemy's session flush event as another means of accomplishing this without resorting to triggers. That has the advantage of being in the python code instead of the DB which is probably easier to maintain.

Since we may have to do this in more tables, it might be worth another look. I think either approach is reasonable, but it's probably easier to maintain a few lines of python than a bunch of triggers. I'll fiddle around with it.

Thanks for looking at what I got this far. Mainly I wanted to present the idea and see if anybody had strong objections to this strategy. I really don't care how we update the timestamp so long as it's updated by the time I spam it. :)

Nerdinacan committed:
Merge branch 'dev' into update_time_trigger

Nerdinacan committed:
Added update_time trigger functions and timestamp fields

Nerdinacan committed:
added timestamp filters to api

Nerdinacan committed:
removed features accidentally copied from another working branch

Nerdinacan committed:
changed triggers to before update, added conditional column adds

Nerdinacan committed:
removed drop column because it was requested that update_time go in the initial db mappings

nsoranzo commented:
@Nerdinacan The `downgrade()` in a migration should still undo the changes made in the `upgrade()` (if feasible), but you also need to update `lib/galaxy/model/mapping.py` with the changes from `upgrade()` since for new Galaxy installations the database is not generated via migrations any more, but directly from the definitions in `mapping.py` .
Please let me know if you have doubts.

Nerdinacan commented:
@nsoranzo I actually am a little confused now. Do you not run the migrations at all on new installations as well? Shouldn't the migrations return the db to the state it was in before the upgrade() ran? If the timestamps were in mappings originally then the state before the upgrade() ran already had timestamps. Isn't the conditional check the correct move? 

And, if the migrations DO NOT run on new installations does that mean the triggers I implemented need to somehow be run at the same time mappings does in a new installation?

martenson commented:
@Nerdinacan we do not run migration on clean new db since https://github.com/galaxyproject/galaxy/pull/8300

Nerdinacan commented:
@martenson Ok cool. So I DO want those triggers installed at the same time mappings goes in then. Because otherwise new installations will not receive them.

martenson commented:
@Nerdinacan That is my understanding also.

Nerdinacan commented:
@martenson @nsoranzo Ok I'll try to figure out where best to execute the installation.

Nerdinacan committed:
restored drop-timestamps because this migration only operates on existing databases

Nerdinacan committed:
Referenced galaxy aliased orm.now

Nerdinacan committed:
Merge branch 'dev' into update_time_trigger

Nerdinacan committed:
sMerge branch 'dev' into update_time_trigger

Nerdinacan committed:
Moved triggers into main model init re:8300

Nerdinacan committed:
Merge branch 'dev' into update_time_trigger

Nerdinacan committed:
removed unnecessary criteria in updatetime trigger

Nerdinacan committed:
added check to drop existing triggers upon installation, appears to affect unit tests

Nerdinacan committed:
Merge branch 'dev' into update_time_trigger

Nerdinacan committed:
fixed trigger return values, simplified pg triggers

Nerdinacan commented:
@nsoranzo @martenson  I think this is actually ready now. Finally!

It's been through a couple of iterations: mostly trying to understand the proper return values for the various types of triggers and how that impacted writes.

I moved the trigger generation code inside the model directory and install it when the rest of the database is created. Hopefully i did that properly. I also added some structure around the creation of the triggers so people can reuse the SQL generation as desired.

It's a pretty messy history for 6 file changes. Let me know if you'd like me to give you a cleaner commit to look at.

Nerdinacan committed:
Merge branch 'dev' into update_time_trigger

Nerdinacan committed:
Merge dev to update migration verion number for trigger migration

Nerdinacan committed:
update_time migration version number bump to fit dev

mvdbeek commented on lib/galaxy/model/migrate/versions/0162_add_content_update_time.py:
```diff
@@ -0,0 +1,47 @@
+"""
+Adds timestamps to hdac table. Adds triggers to dataset, hda, hdac tables
```
```suggestion
Adds timestamps to hdca table. Adds triggers to dataset, hda, hdca tables.
```

mvdbeek reviewed (approved):
Looks good to me, the current version number right now would be 164. If you think this is good to go can you update the migration and remove the [WIP] in the PR title ?

dannon committed:
Update migration version number, fix typo
